core/domain/expression_parser.py

```diff
@@ -0,0 +1,597 @@
+    """Class representing the multiplication operator node."""
+
+    def __init__(self, left, right):
+        """Initializes an MultiplicationOperatorNode object.
+
+        Args:
+            left: Node. Left child of the operator.
+            right: Node. Right child of the operator.
+        """
+        super(MultiplicationOperatorNode, self).__init__([left, right])
+
+
+class DivisionOperatorNode(Node):
+    """Class representing the division operator node."""
+
+    def __init__(self, left, right):
+        """Initializes an DivisionOperatorNode object.
+
+        Args:
+            left: Node. Left child of the operator.
+            right: Node. Right child of the operator.
+        """
+        super(DivisionOperatorNode, self).__init__([left, right])
+
+
+class PowerOperatorNode(Node):
+    """Class representing the power operator node."""
+
+    def __init__(self, left, right):
+        """Initializes an PowerOperatorNode object.
+
+        Args:
+            left: Node. Left child of the operator.
+            right: Node. Right child of the operator.
+        """
+        super(PowerOperatorNode, self).__init__([left, right])
+
+
+class IdentifierNode(Node):
+    """Class representing the identifier node. An identifier could be a single
+    latin letter (uppercase/lowercase) or a greek letter represented by the
+    symbol name.
+    """
+
+    def __init__(self, token):
+        """Initializes an IdentifierNode object.
+
+        Args:
+            token: Token. The token representing the identifier.
+        """
+        self.token = token
+        super(IdentifierNode, self).__init__([])
+
+
+class NumberNode(Node):
+    """Class representing the number node."""
+
+    def __init__(self, token):
+        """Initializes a NumberNode object.
+
+        Args:
+            token: Token. The token representing a real number.
+        """
+        self.token = token
+        super(NumberNode, self).__init__([])
+
+
+class UnaryFunctionNode(Node):
+    """Class representing the function node. The functions represented by this
+    class must have exactly one parameter.
+    """
+
+    def __init__(self, token, child):
+        """Initializes a UnaryFunctionNode object.
+
+        Args:
+            token: Token. The token representing the math function.
+            child: Node. The parameter of the function.
+        """
+        self.token = token
+        super(UnaryFunctionNode, self).__init__([child])
+
+
+class Parser(python_utils.OBJECT):
+    """Class representing the math expression parser.
+    Implements a greedy, recursive-descent parser that tries to consume
+    as many tokens as possible while obeying the grammar.
+    More info about recursive-descent parsers:
+    https://en.wikipedia.org/wiki/Recursive_descent_parser
+    """
+
+    def __init__(self):
+        """Initializes the Parser object."""
+        # Stores the index of the next token to be parsed.
+        self.next_token_index = 0
```

Comment:
I've made the `next_token_index` private and added relevant comments to explain it's lifecycle. I've also made the other functions (except `parse`) private and made relevant modifications to the tests. Since I can't call these methods directly, or rather shouldn't, I've refactored the tests such that the checks are based on the production rules that the functions implement rather than testing the functions themselves. So, these test scenarios would ensure that the function related to the production rule mentioned in the docstring is called and produces the expected output.